interop-testing/src/main/java/io/grpc/testing/integration/TestServiceImpl.java

```diff
@@ -296,10 +315,28 @@ public ResponseDispatcher completeInput() {
     }
 
     /**
+     * Allows the service to cancel the remaining responses.
+     */
+    public synchronized void cancel() {
+      if (cancelled) {
+        throw new IllegalStateException("Dispatcher already cancelled");
+      }
+      chunks.clear();
+      cancelled = true;
+    }
+
+    public synchronized boolean isCancelled() {
+      return cancelled;
+    }
+
+    /**
      * Dispatches the current response chunk to the client. This is only called by the executor. At
      * any time, a given dispatch task should only be registered with the executor once.
      */
     private synchronized void dispatchChunk() {
+      if (cancelled) {
```

Comment:
If `canceled` is no longer part of the `ResponseDispatcher` state, all I can do here is check if the queue is empty, e.g.:

```
    private synchronized void dispatchChunk() {
      // If the service already called cancel(), the queue may be empty.
      if (chunks.isEmpty()) {
        return;
      }
      ...
```

It seems like this has the potential to mask other problems with the code by potentially masking cases where the queue is incorrectly empty at this stage. This doesn't seem to be possible with the current code, but still hesitant about adding this kind of check.